R/map_rd2markdown.R

```diff
@@ -102,7 +103,84 @@ clean_text_whitespace <- function(x) {
   x
 }
 
-
+#' Merge whitespace between text elements
+#'
+#' Within a series of text segments, there should be no separate text elements,
+#' which are spaces only. These should be merged together and to the next 
+#' non-space element if possible.
+#' 
+#' @param x `list` of Rd tags
+#'
+#' @examples
+#' \dontrun{
+#' x <- list(" a ", "`b`", " ", "  ", " c ", block(), "\n", "e   \n", "f")
+#' merge_text_whitespaces(x)
+#' # list(" a ", "`b`", "    c ", block(), "\n", "e   \n", "f")
+#'
+#' x <- list(block(), " a ", "`b`", " c ", " ", block(), "   ",  " d ", "e ", "f", block())
+#' merge_text_whitespaces(x)
+#' # list(block(), "a ", "`b`", " c  ", block(), "    d ", "e ", "f", block())
+#' }
+#'
+#' @keywords internal
+merge_text_whitespaces <- function(x) {
+  ws <- vapply(x, function(y) grepl("^ *$", y), FUN.VALUE = logical(1))
+  ib <- vapply(x, function(y) is_block(y),  FUN.VALUE = logical(1))
+  ws <- ws & !ib
+  
+  seqs <- rle(ws)
+  values <- seqs$values
+  lengths <- seqs$lengths
+  
+  merged <- list()
+  i <- 1
+  while (i <= length(values)) {
+    
+    # Current value was already included
+    if (lengths[i] == 0) {
+      i <- i + 1
+      next()
+    }
+    # Get index corresponding to end of current's value subsequence
+    ws_inds <- sum(lengths[seq(i)])
+    # Get subvector of the same value of ws
+    x_sub <- x[seq(ws_inds - lengths[i] + 1, ws_inds)]
+    
+    if (!values[i]) {
+      merged <- append(merged, x_sub)
+    } else {
+      merged_ws <- paste0(x_sub, collapse = "")
+      if (isFALSE(values[i + 1]) && !is_block(x[[ws_inds + 1]])) {
+        merged_ws <- paste0(merged_ws, x[[ws_inds + 1]], collapse = "")
+        # We already include one of the next FALSE, therefore we reduce
+        # corresponding lengths value by 1 and increment current one
+        lengths[i + 1] <- lengths[i + 1] - 1
+        lengths[i] <- lengths[i] + 1
+        merged <- append(merged, list(merged_ws))
+        # If next element is a block, we will try to append to the previous.
+      } else if (isFALSE(values[i - 1])) {
+        # Previous value was FALSE, therefore we can use last value in
+        # merged to append to unless it is block.
+        if (!is_block(merged[[length(merged)]])) {
+          merged_ws <- paste0(merged[[length(merged)]], merged_ws, collapse = "")
+          merged[[length(merged)]] <- merged_ws
+        } else {
+        # Previous value was a block. Merge standalone space TEXT
+          merged_ws <- paste0(x_sub, collapse = "")
+          merged <- append(merged, list(merged_ws))
+        }
+      } else {
+        # Entire x consists of ws only
+        merged_ws <- paste0(x_sub, collapse = "")
+        merged <- append(merged, merged_ws)
+      }
+    }
+    
+    i <- i + 1
+  }
+  
+  merged
+}
```

Comment:
This feels more complicated than I think it needs to be. Instead, we can figure out which bits need to be merged, then paste them together instead of scanning through with many different conditional behaviors. 

**Note:** This isn't a drop-in replacement. Instead of `["  ", "a", "  ", "b", "  "] => ["  a", "  b  "]` as it is now, this will produce `["  a  ", "b  "]`. This only affects the `merge_text_whitespaces` tests, though.

```suggestion
merge_text_whitespaces <- function(x) {
  if (length(x) == 0) return(x)

  # assumes length x > 0
  spaces <- grepl("^ *$", x, perl = TRUE)
  is_trailing <- cumsum(!spaces) > 1  # after at least 1 non-space character
  blocks <- vlapply(x, is_block)
  groups <- cumsum(!spaces & is_trailing | blocks)

  # collapse non-block groups, maintain first-in-group's attributes
  unname(lapply(split(x, groups), function(group) {
    if (is_block(group[[1]])) {
      group[[1]]
    } else  {
      atts <- attributes(group[[1]])
      result <- paste(group, collapse = "")
      attributes(result) <- atts
      result
    }
  }))
}
```

Reply:
Wow, this is much simpler! I feel a little bit miserable now :P 

There are a couple of bugs, however (for instance if the string starts with a block or pearl matching). I will fix them in your version and prompt re-review.

Reply:
I have sligthly refactored your approach, by adding another split logic. The problem with blocks was that they are not only non-space characters but also characters to which we cannot append. That's why it would be really hard to follow your strategy to append always after non-block group. It was leading to skipping first non-space character after block and all space character after block resulting in edge cases like `list(block(), " ", "      ", block())` -> `list(block(), block())` which is something we want to avoid. That's why there were so many edge cases in my original solution. Therefore I simply combined both solutions and removed the problem by doing additional split on each block. Now each sub list is treated (using your brilliant idea) as there were no blocks and everything is merged as expected. Then the blocks and list are pasted back do get the desired result.  I think this is a good compromise.

Reply:
Ah, I see. This might keep the simplicity but resolve these cases:

```r
merge_text_spaces <- function(x) {
  if (length(x) == 0) return(x)

  # assumes length x > 0
  spaces <- grepl("^ *$", x)
  blocks <- vlapply(x, is_block)
  lag_blocks <- c(FALSE, blocks[-length(x)])
  is_trailing_after_block <- (cumsum(!spaces) - cumsum(cumsum(!spaces) * blocks)) > 1
  groups <- cumsum(!spaces & is_trailing_after_block | blocks | lag_blocks)

  # collapse non-block groups, maintain first-in-group's attributes
  unname(lapply(split(x, groups), function(group) {
    if (is_block(group[[1]])) {
      group[[1]]
    } else  {
      atts <- attributes(group[[1]])
      result <- paste(group, collapse = "")
      attributes(result) <- atts
      result
    }
  }))
}
```

Reply:
`is_trailing_after_block <- (cumsum(!spaces) - cumsum(cumsum(!spaces) * blocks)) > 1` although it's a really smart neat solution, is it simpler? There is a lot of happening in this line and might be hard to unpack for someone reading this code in the future (and this is an open source). To be honest I'm leaning towards just splitting the list on each block and treating strings as if there were no blocks and merging afterwards. I think might be simpler to work around this in the future. What's your opinion on this one?